src/main/scala/is/hail/nativecode/NativeCode.java

```diff
@@ -1,42 +1,151 @@
 package is.hail.nativecode;
 
 import java.io.*;
+import java.util.*;
+import java.util.regex.*;
+import java.util.zip.*;
+import java.net.URL;
+import java.nio.file.*;
+import com.sun.jna.*;
+import org.apache.log4j.*;
 
-class NativeCode {
+public class NativeCode {
+  private static String includeDir;
+  private static String hailName;
+  
   static {
+    hailName = "hail_abi_v9";
     try {
-      File file = File.createTempFile("libhail", ".lib");
-      ClassLoader loader = NativeCode.class.getClassLoader();
-      InputStream s = null;
-      String osName = System.getProperty("os.name");
-      if (osName.equals("Linux") || osName.equals("linux")) {
-        s = loader.getResourceAsStream("linux-x86-64/libhail.so");
+      // libboot.so has native methods to call dlopen/dlclose
+      String libBoot = libToLocalFile("libboot");
+      System.load(libBoot);
+      if (isLinux()) {
```

Comment:
I don't like where this is going.  This is what I want to see for the build system:

1. The static (libhail, etc.) and dynamic (generated) Makefiles setup should be the same.  It might make sense to have a common setup in a Makefile fragment you include in the installation/distribution.  I don't see a scenario where looking up e.g. CXX in the environment from C++/Scala makes sense.

2. Local builds (e.g. for development) and local installations* should rely on the pre-installed C++ compiler in the usual way (with warnings, etc.)  In particularly, they shouldn't rely on prebuilt (which is a mistake in the current setup).  Prebuilt is intended to store the objects for the purpose of deployment when we didn't have access to OSX machines.

3. Compatibility with the TeamCity CI is not a goal here.  It can be changed and will be deprecated soon.  Here are instructions for changing the CI image: https://github.com/hail-is/hail-infrastructure/blob/master/how-to-modify-the-CI-worker-image.md.  Or coordinate wth @danking who is moving the deployment over to the new CI.

4. You convinced me we need to deploy with a compiler (and other C++ dependencies).  Let's do that.  It completely obviates these ABI considerations.  We will need separate Linux and OSX distributions.  We've ordered an OSX computer to automate the OSX distribution.  I will check if we have a delivery date yet.  Again, coordinate with @danking.  We can use prebuilt until the OSX build machine comes online.

5. *We don't really have an "install" operation currently, users doing their own install just copy the jar/Python zip to the right place.  As Hail requires more artifacts now (e.g. Makefile snippets, header files, etc.), I think a formal installation makes sense.

6. In particular, the header files shouldn't be in the JAR.  They should be in a standard place in the installation, e.g. $HAIL_HOME/include.

While you're getting the distribution set up, it makes sense to get the C++ decoder in in parallel under control of a feature flag (off by default) for the purposes of development.  We should still test the code with the feature flag on.  You can control the environment for that through Docker in the new test setup.

Let me know if this picture isn't clear.

Reply:
> get the code committed, so that people could start playing with it

There are two sets of people here: people on the Hail team and Hail users.  We should focus first on getting the code into the hands of people on the team so they can start understanding/building on it, and that shouldn't require solving all the distribution problems.  I think that should happen first, or at least in parallel.

 > FWIW, if you do a build then it overwrites the libraries in prebuilt, so you in that case everything does in fact match up, it's just confusing because "prebuilt" isn't really prebuilt.

We shouldn't overwrite prebuilt.  Prebuilt was created exactly to avoid overwriting stuff during the build.  We should have a command to update prebuilt from the build.

 > Are you thinking the feature flag would be an environment variable, e.g. HAIL_ENABLE_CPP_CODEGEN or something similar ?

Yeah, that'd be great.

Reply:
>We should focus first on getting the code into the hands of people on the team

Agreed. The difficulty is doing that in a non-disruptive way while the policy so far
is to always/only merge changes directly onto master, which then goes out to 
non-developer users.

Absent that, I would put it on a non-master branch with use-whatever-CXX-you've-got.
I'm not particularly advocating for a change of process, but this is a case where
leaving out staging and an old-style QA process raises the stakes.

Having said that, I guess it works as long as a) I make it good enough to get through
CI testing (which was an issue before I went on vacation, but I've done it at least once now),
and b) the actual native PackDecoder is disabled by default (though that isn't in this PR, it will
be in the followup).